src/twisted/internet/asyncioreactor.py

```diff
@@ -9,42 +9,56 @@
 
 import errno
 
+from typing import Dict, Optional, Type, cast
+
 from zope.interface import implementer
 
 from twisted.logger import Logger
 from twisted.internet.posixbase import (PosixReactorBase, _NO_FILEDESC,
                                         _ContinuousPolling)
 from twisted.python.log import callWithLogger
 from twisted.python.runtime import seconds as runtimeSeconds
+from twisted.internet.abstract import FileDescriptor
 from twisted.internet.interfaces import IReactorFDSet
 
-try:
-    from asyncio import get_event_loop
-except ImportError:
-    raise ImportError("Requires asyncio.")
-
-# As per ImportError above, this module is never imported on python 2, but
-# pyflakes still runs on python 2, so let's tell it where the errors come from.
-from builtins import PermissionError, BrokenPipeError
+from asyncio import get_event_loop, AbstractEventLoop, SelectorEventLoop
 
 
 
 @implementer(IReactorFDSet)
 class AsyncioSelectorReactor(PosixReactorBase):
     """
     Reactor running on top of L{asyncio.SelectorEventLoop}.
+
+    On POSIX platforms, the default event loop is
+    L{asyncio.SelectorEventLoop}.
+    On Windows, the default event loop on Python 3.7 and older
+    is L{asyncio.WindowsSelectorEventLoop}, but on Python 3.8 and newer
+    the default event loop is L{asyncio.WindowsProactorEventLoop} which
+    is incompatible with L{AsyncioSelectorReactor}.
+    Applications that use L{AsyncioSelectorReactor} on Windows
+    with Python 3.8+ must call
+    C{asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())}
+    before instantiating and running L{AsyncioSelectorReactor}.
     """
     _asyncClosed = False
     _log = Logger()
 
-    def __init__(self, eventloop=None):
+    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
+        if not eventloop:
+            _eventloop = get_event_loop()  # type: AbstractEventLoop
+            eventloop = cast(Optional[SelectorEventLoop], _eventloop)
 
-        if eventloop is None:
-            eventloop = get_event_loop()
+        # On Python 3.8+, asyncio.get_event_loop() on
+        # Windows was changed to return a ProactorEventLoop
+        if not isinstance(eventloop, SelectorEventLoop):
+            raise ValueError(
+                "eventloop is: {} which is not a subclass of "
+                "SelectorEventLoop".format(eventloop))
```

Comment:
```suggestion
            raise ValueError(
                "SelectorEventLoop required, instead got:"
                + " {}".format(eventloop))
```
It's not obvious to me that the variable name should be referenced, especially since the value might come from inside this function as opposed to the parameter.

It doesn't seem that there is any Twisted policy on this, but for the record, my preference is to have specific exceptions from the library rather than general stdlib exceptions that are harder to achieve uniqueness of.  Picking apart exceptions based on their message is a pain when trying to catch only specific things.

Reply:
OK, incorporated.